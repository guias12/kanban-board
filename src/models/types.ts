export interface ICard {
  id: number;
  text: string;
}

export interface IList {
  title: string;
  id: number;
  cards: ICard[];
}
