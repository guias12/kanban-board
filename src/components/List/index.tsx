import React, { useState, KeyboardEvent } from 'react';
import Card from '../Card';
import { ICard } from '../../models/types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faTimes } from '@fortawesome/free-solid-svg-icons';
import { Droppable } from 'react-beautiful-dnd';
import { useDispatch } from 'react-redux';
import { addCard } from '../../reducers/listSlice';
import './styles.scoped.scss';
interface IListProps {
  title: string;
  id: number;
  cards: ICard[];
  searchTerm: string;
}

const List = ({ title, id, cards, searchTerm }: IListProps) => {
  const dispatch = useDispatch();
  const [creatingCard, setCreatingCard] = useState(false);
  const [newCardText, setNewCardText] = useState('');

  const toggleCardCreation = (status: boolean): void => {
    setCreatingCard(status);

    if (!status) {
      setNewCardText('');
    }
  };

  const handleInputKeyDown = (event: KeyboardEvent<HTMLInputElement>): void => {
    if (event.key === 'Enter') {
      handleAddCard();
    }
  };

  const handleAddCard = (): void => {
    if (newCardText) {
      dispatch(
        addCard({
          text: newCardText,
          listId: id,
        })
      );
    }
    setCreatingCard(false);
    setNewCardText('');
  };
  return (
    <Droppable droppableId={`${id}`}>
      {(provided) => (
        <div
          {...provided.droppableProps}
          ref={provided.innerRef}
          className="list-container"
        >
          <h3>{title}</h3>
          {cards.map((card, index) => {
            if (searchTerm !== '') {
              if (card.text.toLowerCase().includes(searchTerm.toLowerCase())) {
                return (
                  <Card
                    card={card}
                    listId={id}
                    id={card.id}
                    key={card.id}
                    index={index}
                  />
                );
              } else {
                return null;
              }
            } else {
              return (
                <Card
                  card={card}
                  listId={id}
                  id={card.id}
                  key={card.id}
                  index={index}
                />
              );
            }
          })}
          {creatingCard ? (
            <div className="new-card-container">
              <input
                type="text"
                name="new-card-text"
                id="new-card-text"
                className="new-card-text"
                placeholder="Add new card text"
                value={newCardText}
                onChange={(e) => setNewCardText(e.target.value)}
                onKeyDown={handleInputKeyDown}
                autoFocus
                data-testid="list-create-input"
              />
              <div className="new-card-buttons">
                <button
                  className="add-card-btn"
                  onClick={handleAddCard}
                  data-testid="list-save-button"
                >
                  Add card
                </button>
                <button
                  className="cancel-add-btn"
                  onClick={() => toggleCardCreation(false)}
                  data-testid="list-cancel-button"
                >
                  <FontAwesomeIcon icon={faTimes} />
                </button>
              </div>
            </div>
          ) : (
            <button
              className="list-add-card-btn"
              type="button"
              onClick={() => toggleCardCreation(true)}
            >
              <FontAwesomeIcon icon={faPlus} />
              Create new card
            </button>
          )}
          {provided.placeholder}
        </div>
      )}
    </Droppable>
  );
};

export default List;
