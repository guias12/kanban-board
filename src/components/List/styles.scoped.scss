@import '../../components/styles/global.scss';
.list-container {
  background: #ebecf0;
  box-shadow: rgba(50, 50, 93, 0.25) 0px 50px 100px -20px,
    rgba(0, 0, 0, 0.3) 0px 30px 60px -30px;
  border-radius: 8px;
  width: 450px;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 5px 15px;
  max-height: 75%;
  overflow-y: auto;
  padding-bottom: 20px;
}

.list-add-card-btn {
  width: 90%;
  padding: 10px 15px;
  cursor: pointer;
  color: white;
  font-size: 1em;
  border: none;
  margin-top: 10px;
  background: var(--primary);
  transition: 0.2s;
}

.list-add-card-btn > svg {
  margin-right: 5px;
}

.list-add-card-btn:hover {
  background: var(--secondary);
}

.new-card-container {
  min-height: 40px;
  margin-top: 15px;
  width: 90%;
  display: flex;
  flex-direction: column;
}

.new-card-text {
  padding: 10px;
  border: none;
  background: white;
  border-radius: 4px;
  font-size: 0.8em;
}

.new-card-buttons {
  margin-top: 5px;

  > button {
    margin-top: 5px;
    margin-right: 10px;
    cursor: pointer;
    border: none;
    border-radius: 4px;
    box-shadow: rgba(50, 50, 93, 0.25) 0px 50px 100px -20px,
      rgba(0, 0, 0, 0.3) 0px 30px 60px -30px;
    color: white;
    transition: 0.2s;
  }
}

.cancel-add-btn {
  background: var(--danger);
  padding: 8px 13px;
}
.cancel-add-btn:hover {
  background: var(--danger-light);
}

.add-card-btn {
  background: var(--primary);
  padding: 8px 10px;
}
.add-card-btn:hover {
  background: var(--secondary);
}
