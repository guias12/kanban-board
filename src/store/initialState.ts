import { IInitialState } from '../models/types';
const getInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
});

const getEhancedTestInitialState = (): IInitialState => ({
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'First ToDo card',
        },
        {
          id: 1,
          text: 'Second ToDo card',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [
        {
          id: 2,
          text: 'first Doing card',
        },
        {
          id: 3,
          text: 'second Doing card',
        },
        {
          id: 4,
          text: 'third Doing card',
        },
      ],
    },
    {
      title: 'Done',
      id: 2,
      cards: [
        {
          id: 5,
          text: 'first Done card',
        },
        {
          id: 6,
          text: 'second Done card',
        },
      ],
    },
  ],
});

export { getInitialState, getEhancedTestInitialState };
