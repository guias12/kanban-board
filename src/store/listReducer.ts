import { IInitialState } from '../models/types';
import { AnyAction } from 'redux';
import { CONSTANTS } from '../actions';

const initialState: IInitialState = {
  searchTerm: '',
  list: [
    {
      title: 'To Do',
      id: 0,
      cards: [
        {
          id: 0,
          text: 'Test kanban board',
        },
      ],
    },
    {
      title: 'Doing',
      id: 1,
      cards: [],
    },
    {
      title: 'Done',
      id: 2,
      cards: [],
    },
  ],
};
let nextAvailableCardId = 1;

const listReducer = (state = initialState, action: AnyAction) => {
  const stateClone: IInitialState = { ...state };

  switch (action.type) {
    case CONSTANTS.ADD_CARD:
      const newStateList = stateClone.list.map((list) => {
        if (list.id === action.payload.listId) {
          const newCardId = nextAvailableCardId;
          const newCard = {
            text: action.payload.text,
            id: newCardId,
          };
          nextAvailableCardId += 1;

          return {
            ...list,
            cards: [...list.cards, newCard],
          };
        } else {
          return list;
        }
      });

      stateClone.list = newStateList;
      state.list = newStateList;
      return stateClone;
    case CONSTANTS.CHANGE_CARD_TEXT:
      const { listId, cardId, text } = action.payload;
      let currentList = stateClone.list.find((list) => list.id === listId);
      const currentListIndex = stateClone.list.findIndex(
        (list) => list.id === listId
      );
      const edditingCardIndex = currentList?.cards.findIndex(
        (card) => card.id === cardId
      );

      if (
        (edditingCardIndex || edditingCardIndex === 0) &&
        (currentList || currentList === 0)
      ) {
        let newCards = [...currentList.cards];
        newCards[edditingCardIndex].text = text;
        currentList = {
          ...currentList,
          cards: [...newCards],
        };

        stateClone.list[currentListIndex] = currentList;
        return stateClone;
      }

      return state;
    case CONSTANTS.DRAG_CARD:
      const {
        sourceListId,
        destinationListId,
        sourceCardIndex,
        destinationCardIndex,
      } = action.payload;

      if (sourceListId === destinationListId) {
        const list = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = list?.cards.splice(sourceCardIndex, 1);

        if (card) {
          list?.cards.splice(destinationCardIndex, 0, ...card);
        }
      } else {
        const sourceList = stateClone.list.find(
          (list) => sourceListId === String(list.id)
        );
        const card = sourceList?.cards.splice(sourceCardIndex, 1);
        const destinationList = stateClone.list.find(
          (list) => destinationListId === String(list.id)
        );
        if (card) {
          destinationList?.cards.splice(destinationCardIndex, 0, ...card);
        }
      }

      return stateClone;
    case CONSTANTS.FILTER_CARD:
      const { searchTerm } = action.payload;
      stateClone.searchTerm = searchTerm;

      return stateClone;
    default:
      return state;
  }
};

export default listReducer;
