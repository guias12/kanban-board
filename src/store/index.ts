import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useSelector } from 'react-redux';
import listReducer from './listSlice';

const store = configureStore({
  reducer: {
    list: listReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export type AppDispatch = typeof store.dispatch;

export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
