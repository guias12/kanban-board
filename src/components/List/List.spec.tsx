import React from 'react';
import List from './index';
import { Droppable } from 'react-beautiful-dnd';
import { DragDropContext } from 'react-beautiful-dnd';
import { ICard } from '../../models/types';
import { render, screen } from '@testing-library/react';
import configureMockStore, { MockStoreEnhanced } from 'redux-mock-store';
import { getInitialState } from '../../store/initialState';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';
import fireEvent from '@testing-library/user-event';

describe('List component', () => {
  let store: MockStoreEnhanced<unknown, {}>;
  const cards: ICard[] = [
    { id: 0, text: 'abc' },
    { id: 1, text: '123' },
    { id: 2, text: 'lorem ipsum' },
  ];

  const setup = (searchTerm: string = ''): JSX.Element => {
    const initialState = getInitialState();
    store = configureMockStore()({ list: initialState });

    return (
      <Provider store={store}>
        <DragDropContext onDragEnd={() => {}}>
          <List
            title="Test list"
            id={0}
            cards={cards}
            searchTerm={searchTerm}
          />
        </DragDropContext>
      </Provider>
    );
  };

  it('Should render List component correctly', () => {
    render(setup());
    const listTitle = screen.getByText(/Test list/i);

    expect(listTitle).toBeInTheDocument();
  });

  it('Should allow user to create new card on button click', () => {
    render(setup());
    const newCardButton = screen.getByText(/Create new card/i);
    fireEvent.click(newCardButton);

    const newCardInput = screen.getByTestId('list-create-input');

    expect(newCardInput).toBeInTheDocument();
  });

  it('Should allow user to cancel new card creation on button click', () => {
    render(setup());
    const newCardButton = screen.getByText(/Create new card/i);
    fireEvent.click(newCardButton);

    const cancelCreationButton = screen.getByTestId('list-cancel-button');
    fireEvent.click(cancelCreationButton);

    expect(cancelCreationButton).not.toBeInTheDocument();
  });

  it('Should render Cards filtered in a correct way', () => {
    render(setup(cards[0].text));
    const filteredCard = screen.getByText(cards[0].text);

    expect(filteredCard).toBeInTheDocument();

    const invalidCard = screen.queryByText(cards[1].text);

    expect(invalidCard).not.toBeInTheDocument();
  });

  it('Should not render any Cards if filtering invalid words', () => {
    render(setup('this is not a valid words combination for filtering'));
    const card0 = screen.queryByText(cards[0].text);
    const card1 = screen.queryByText(cards[1].text);
    const card2 = screen.queryByText(cards[2].text);

    expect(card0).not.toBeInTheDocument();
    expect(card1).not.toBeInTheDocument();
    expect(card2).not.toBeInTheDocument();
  });

  it('Should match List snapshot', () => {
    const component = create(setup());
    const componentJson = component.toJSON();

    expect(componentJson).toMatchSnapshot();
  });
});
