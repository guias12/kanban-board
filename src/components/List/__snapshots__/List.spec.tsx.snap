// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Should match List snapshot 1`] = `
<div
  className="list-container"
  data-rbd-droppable-context-id="5"
  data-rbd-droppable-id="0"
>
  <h3>
    Test list
  </h3>
  <div
    aria-describedby="rbd-hidden-text-5-hidden-text-20"
    className="card-container"
    data-rbd-drag-handle-context-id="5"
    data-rbd-drag-handle-draggable-id="0"
    data-rbd-draggable-context-id="5"
    data-rbd-draggable-id="0"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={0}
  >
    <span
      className="card-text"
      data-testid="card-text"
      onClick={[Function]}
    >
      abc
    </span>
  </div>
  <div
    aria-describedby="rbd-hidden-text-5-hidden-text-20"
    className="card-container"
    data-rbd-drag-handle-context-id="5"
    data-rbd-drag-handle-draggable-id="1"
    data-rbd-draggable-context-id="5"
    data-rbd-draggable-id="1"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={0}
  >
    <span
      className="card-text"
      data-testid="card-text"
      onClick={[Function]}
    >
      123
    </span>
  </div>
  <div
    aria-describedby="rbd-hidden-text-5-hidden-text-20"
    className="card-container"
    data-rbd-drag-handle-context-id="5"
    data-rbd-drag-handle-draggable-id="2"
    data-rbd-draggable-context-id="5"
    data-rbd-draggable-id="2"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={0}
  >
    <span
      className="card-text"
      data-testid="card-text"
      onClick={[Function]}
    >
      lorem ipsum
    </span>
  </div>
  <button
    className="list-add-card-btn"
    onClick={[Function]}
    type="button"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-plus "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
    Create new card
  </button>
</div>
`;
